File: validate_test.go
// MIT License
//
// Copyright (c) 2026 sparetimecoders
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package spec

import (
	"encoding/json"
	"os"
	"strings"
	"testing"
)

func loadValidateFixtures(t *testing.T) validateFixtures {
	t.Helper()
	data, err := os.ReadFile("testdata/validate.json")
	if err != nil {
		t.Fatalf("failed to read validate fixtures: %v", err)
	}
	var f validateFixtures
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("failed to parse validate fixtures: %v", err)
	}
	return f
}

func TestValidate(t *testing.T) {
	for _, tc := range loadValidateFixtures(t).Single {
		t.Run(tc.Name, func(t *testing.T) {
			err := Validate(tc.Topology)
			if tc.ExpectedError == nil {
				if err != nil {
					t.Errorf("expected no error, got: %v", err)
				}
			} else {
				if err == nil || !strings.Contains(err.Error(), *tc.ExpectedError) {
					t.Errorf("expected error containing %q, got: %v", *tc.ExpectedError, err)
				}
			}
		})
	}
}

func TestValidateTopologies(t *testing.T) {
	for _, tc := range loadValidateFixtures(t).Cross {
		t.Run(tc.Name, func(t *testing.T) {
			err := ValidateTopologies(tc.Topologies)
			if tc.ExpectedError == nil {
				if err != nil {
					t.Errorf("expected no error, got: %v", err)
				}
			} else {
				if err == nil || !strings.Contains(err.Error(), *tc.ExpectedError) {
					t.Errorf("expected error containing %q, got: %v", *tc.ExpectedError, err)
				}
			}
		})
	}
}

File: nodejs-spec/__tests__/validate.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import { validate, validateTopologies } from "../src/validate.js";
import type { Topology } from "../src/types.js";

const fixturesPath = resolve(
  import.meta.dirname,
  "../../testdata/validate.json",
);
const fixtures = JSON.parse(readFileSync(fixturesPath, "utf-8"));

describe("validate", () => {
  describe("single topology", () => {
    for (const tc of fixtures.single) {
      it(tc.name, () => {
        const result = validate(tc.topology as Topology);
        if (tc.expectedError === null) {
          expect(result).toBeNull();
        } else {
          expect(result).not.toBeNull();
          expect(result).toContain(tc.expectedError);
        }
      });
    }
  });

  describe("cross-topology validation", () => {
    for (const tc of fixtures.cross) {
      it(tc.name, () => {
        const result = validateTopologies(tc.topologies as Topology[]);
        if (tc.expectedError === null) {
          expect(result).toBeNull();
        } else {
          expect(result).not.toBeNull();
          expect(result).toContain(tc.expectedError);
        }
      });
    }
  });
});
